### PR TITLE
drain mysql protocol connections to vtgate at shutdown

### PR DESCRIPTION
**Pull request by tpetr**
This PR adds support for draining mysql protocol connections from the vtgate when it terminates.

### CODE REVIEW

File: go/mysql/server.go [hunk 1]
```diff
@@ -114,6 +120,11 @@ type Listener struct {
 
 	// Incrementing ID for connection id.
 	connectionID uint32
+
+	shutdownOnce      sync.Once
```
Comment (tpetr) [inline go/mysql/server.go]: the `sync.Once` may be overkill here

File: go/mysql/server.go [hunk 2]
```diff
@@ -147,13 +160,19 @@ func (l *Listener) Accept() {
 			return
 		}
 
+		if l.shutdown {
+			conn.Close()
```
Comment (tpetr) [inline go/mysql/server.go]: not sure if I should add some debug logging here or what

File: go/mysql/server.go [hunk 3]
```diff
@@ -294,10 +314,28 @@ func (l *Listener) handle(conn net.Conn, connectionID uint32, acceptTime time.Ti
...
-		data, err := c.readEphemeralPacket()
-		if err != nil {
+		go func() {
+			data, err := c.readEphemeralPacket()
+			if err != nil {
+				errorCh <- err
+			} else {
+				packetCh <- data
+			}
+		}()
```
Comment (tpetr) [inline go/mysql/server.go]: The idea here is to make it so that `shutdownCh` can preempt `readEphemeralPacket()`. This approach feels a little gross and I'm fairly new to Go, so any pointers here are appreciated.

File: go/mysql/server.go [hunk 4]
```diff
@@ -407,6 +447,19 @@ func (l *Listener) Close() {
...
+// Shutdown starts closing idle connections.
+func (l *Listener) Shutdown() {
+	l.shutdownOnce.Do(func() {
+		l.shutdown = true
+		close(l.shutdownCh)
+	})
+}
+
+// WaitForConnections blocks until there are no active connections
+func (l *Listener) WaitForConnections() {
```
Comment (tpetr) [inline go/mysql/server.go]: I could probably name this better

**Comment by tpetr** [comment]
I signed it!

**Comment by tpetr** [comment]
@sougou ready for your review now. I need to finish adding unit tests, but I've manually verified:
- idle connections are immediately closed
- connections with an active query are closed when the query completes
- connections that are in the middle of a transaction are closed when the transaction is complete

File: go/mysql/server.go [hunk 5]
```diff
@@ -294,10 +315,37 @@ func (l *Listener) handle(conn net.Conn, connectionID uint32, acceptTime time.Ti
...
+			data, err := c.readEphemeralPacket()
+
+			if err != nil {
+				errorCh <- err
+				return
+			}
+
+			packetCh <- data
+		}
+	}()
```
Comment (tpetr) [inline go/mysql/server.go]: If it's not obvious, the point of moving `readEphemeralPacket()` into its own goroutine is so that `shutdownCh` can preempt it (since it blocks until a packet has been read or the connection has closed). The `packetCh` and `errorCh` channels feel a little gross to me, but I'm not sure if theres a better way (still somewhat of a Go newbie). I can add a comment explaining all this if it helps.

File: go/vt/vtgate/plugin_mysql_server.go [hunk 6]
```diff
@@ -216,18 +225,31 @@ func newMysqlUnixSocket(address string, authServer mysql.AuthServer, handler mys
...
+		log.Info("Waiting for mysql connections to drain...")
+		wg.Wait()
+		log.Info("All mysql connections drained successfully!")
 	})
 }
 
+func drainListener(l *mysql.Listener, wg *sync.WaitGroup) {
+	wg.Add(1)
+	go func() {
+		l.Shutdown()
```
Comment (sougou) [inline go/vt/vtgate/plugin_mysql_server.go]: This is not required. You can actually delete all the logic involved with Shutdown. Bascially, l.Close() already does all this, which is to stop listening.

File: go/mysql/server.go (hunk 1 above)
Comment (sougou) [reply go/mysql/server.go]: As mentioned in the `drainListener` code, none of this is required. If you just closed the listener, the underlying network code already does this type of thing.

Your existing connections will still be open, which you 'll have to drain.

**Comment by sougou** [comment]
Basically, there's no need to change anything in mysql. All changes could be made at the plugin level.
Here's what I had in mind (may not compile), and you may have to tweak this a bit:

```
var (
  busyLock sync.Mutex
  busyCond = sync.NewCond(&busyLock)
  busy = make(map[*mysql.Conn]bool)
)
```

- Beginning of ComQuery:

```
busyLock.Lock()
busy[c] = true
busyLock.Unlock()

defer func() {
  busyLock.Lock()
  defer busyLock.Unlock()
  busy[c] = session.InTransaction
  if !session.InTransaction {
    // This will wake up the OnTerm function if it's waiting for connections to become non-busy.
    busyCond.Broadcast()
  }
}()
```

- OnTerm:
```
mysqlListener.Close()
busyLock.Lock()
defer busyLock.Unlock()
for {
  var notFinished bool
  for c, busy := range busy {
    if !busy {
      c.Close()
      delete(busy, c)
    }
    notFinished = true
  }
  if notFinished {
    busyCond.Wait()
    continue
  }
  break
}
```

**Comment by tpetr** [comment]
Thanks a lot for the feedback, that is much nicer! A lot of the complexity in my original implementation was due to the fact that I misunderstood `Listener.Close()` -- for some reason I thought it closed all open connections as well.

File: go/vt/vtgate/plugin_mysql_server.go [hunk 7]
```diff
@@ -79,6 +84,10 @@ func (vh *vtgateHandler) ComQuery(c *mysql.Conn, query string, callback func(*sq
 	// FIXME(alainjobart): Add some kind of timeout to the context.
 	ctx := context.Background()
 
+	busyLock.Lock()
+	busy[c] = true
+	busyLock.Unlock()
+
```
Comment (sougou) [inline go/vt/vtgate/plugin_mysql_server.go]: Better to move this section down where the defer is called. That way, there's no risk of someone accidentally adding a return between here and there.

File: go/vt/vtgate/plugin_mysql_server.go [hunk 8]
```diff
@@ -79,6 +84,10 @@ func (vh *vtgateHandler) ComQuery(c *mysql.Conn, query string, callback func(*sq
 	// FIXME(alainjobart): Add some kind of timeout to the context.
 	ctx := context.Background()
 
+	busyLock.Lock()
+	busy[c] = true
```
Comment (demmer) [inline go/vt/vtgate/plugin_mysql_server.go]: It looks like nothing cleans up the record from the map until shutdown is called --you need to also delete it when the connection goes away for normal reasons or we will blow up on memory.

File: go/vt/vtgate/plugin_mysql_server.go [hunk 9]
```diff
@@ -225,6 +243,30 @@ func init() {
...
+		defer busyLock.Unlock()
+		if len(busy) == 0 {
+			log.Info("No mysql connections to drain.")
+			return
+		}
+		log.Infof("Draining %v mysql connection(s)...", len(busy))
+		for {
+			var notFinished bool
+			for c, connIsBusy := range busy {
+				if !connIsBusy {
```
Comment (demmer) [inline go/vt/vtgate/plugin_mysql_server.go]: It seems like there's a race when you close the connection out from under it, since another thread could be about to start a new transaction.

File: go/vt/vtgate/plugin_mysql_server.go (hunk 8 above)
Comment (sougou) [reply go/vt/vtgate/plugin_mysql_server.go]: Oops! Forgot about the delete on `ConnectionClosed`.
The alternate approach of just keeping count would also work. The downside is that we have to return a meaningful error on new requests that will force the client to go elsewhere.

But I agree this may be simpler code. We just have to test and make sure that the error returned will cause the client to go elsewhere.

File: go/vt/vtgate/plugin_mysql_server.go (hunk 9 above)
Comment (sougou) [reply go/vt/vtgate/plugin_mysql_server.go]: The race is harmless: If a transaction (or request) is started after connection close, it will get added back to the map, and the outer loop will just re-close it.

But I like the other suggestion of just tracking count, which will completely eliminate this race.

File: go/vt/vtgate/plugin_mysql_server.go [hunk 10]
```diff
@@ -103,6 +110,16 @@ func (vh *vtgateHandler) ComQuery(c *mysql.Conn, query string, callback func(*sq
...
 	}
+
+	if !session.InTransaction {
+		atomic.AddInt32(&busyConnections, 1)
+	}
+	defer func() {
+		if !session.InTransaction {
+			atomic.AddInt32(&busyConnections, -1)
+		}
+	}()
```
Comment (tpetr) [inline go/vt/vtgate/plugin_mysql_server.go]: I did this semi-complicated logic so that we would consider connections as active during the life of a transaction, but on second thought, maybe we should just allow the vtgate to terminate mid-transaction and ensure that the rollback occurs?

**Comment by tpetr** [comment]
Thanks for the feedback @demmer! I've been flip-flopping -- I like the idea of proactively "pushing" clients away from the terminating vtgate -- but this approach is definitely simpler.

**Comment by sougou** [comment]
LGTM
This is sooo simple that I have to accept it :).

**Comment by demmer** [comment]
Did you consider writing any tests to verify this works?

I stared at the code for a bit and convinced myself that in all the various states of connections / transactions, this _looks_ like it does the right thing, but I would love to see some unit tests to prove it with a few permutations of selects / autocommit DMLs / explicit transactions.

At first glance it looks like you should be able to just add some test cases to the existing `mysql_protocol_test.go`.

I'd also like to see this count exposed as a /debug/vars stat, it'd be nice to print the number periodically (say every 1s) during the shutdown sequence, and I would like the total amount of time that it will wait to drain capped at something reasonable like 10s.

Basically what worries me about this is that it introduces a failure mode where a vtgate is trying to be restarted and if there is a bug in the accounting, could be stuck forever waiting for the miscounted "busy" transaction.
